File: group/group.go
package group

import (
	"fmt"
	"regexp"
	"strconv"
	"sync"
	"time"

	"gopkg.in/tucnak/telebot.v2"

	"github.com/piggy-moe/sticker-captcha-bot/bot"
	"github.com/piggy-moe/sticker-captcha-bot/i18n"
	"github.com/piggy-moe/sticker-captcha-bot/log"
	"github.com/piggy-moe/sticker-captcha-bot/redis"
)

type role string

const (
	roleNone   role = "none"
	roleMember role = "member"
	roleAdmin  role = "admin"
)

type action string

const (
	actionKick action = "kick"
	actionMute action = "mute"
	actionBan  action = "ban"
)

var index sync.Map

func Get(id int64) *Group {
	n := New(id)
	g, loaded := index.LoadOrStore(id, n)
	if loaded {
		log.Tracef("group.get(%d): ok loaded", id)
		return g.(*Group)
	}
	log.Tracef("group.get(%d): ok stored", id)
	go n.start()
	return n
}

type Update struct {
	msg      *telebot.Message
	failUser *telebot.User
}

func NewUpdate(upd *telebot.Update) *Update {
	return &Update{msg: upd.Message}
}

type Group struct {
	id      int64
	Updates chan *Update
}

func New(id int64) *Group {
	return &Group{
		id:      id,
		Updates: make(chan *Update, 100),
	}
}

func (g *Group) start() {
	for upd := range g.Updates {
		m := upd.msg
		if upd.failUser != nil && g.existsf("enabled") {
			if g.existsf("user:%d:pending", upd.failUser.ID) {
				g.onFail(upd.msg, upd.failUser)
			}
			continue
		}
		for _, u := range m.UsersJoined {
			g.delf("user:%d:role", u.ID)
		}
		if g.tryHandleVerification(m) {
			continue
		}
		g.handleCommand(m)
	}
}

func (g *Group) tryHandleVerification(m *telebot.Message) bool {
	if !g.existsf("enabled") {
		return false
	}
	if m.IsService() {
		for _, u := range m.UsersJoined {
			g.onJoin(m, &u)
		}
		return true
	}
	if g.existsf("user:%d:pending", m.Sender.ID) {
		if m.Sticker != nil {
			g.onPass(m, m.Sender)
		} else {
			g.delete(m.ID)
		}
		return true
	}
	return false
}

func (g *Group) onJoin(m *telebot.Message, u *telebot.User) {
	log.Tracef("group(%d).onjoin(mid=%d, uid=%d)", g.id, m.ID, u.ID)
	g.setf("user:%d:pending", "true", 0, u.ID)
	go func(g *Group, m *telebot.Message, u *telebot.User) {
		h := g.send(g.render(g.getOnJoinTemplate(), u), m.ID)
		time.Sleep(g.getTimeout())
		g.Updates <- &Update{
			msg:      m,
			failUser: u,
		}
		if g.existsf("verbose") {
			return
		}
		g.delete(h)
	}(g, m, u)
}

func (g *Group) onPass(m *telebot.Message, u *telebot.User) {
	log.Tracef("group(%d).onpass(mid=%d, uid=%d)", g.id, m.ID, u.ID)
	g.delf("user:%d:pending", u.ID)

	if g.existsf("quiet") {
		go g.delete(m.ID)
		return
	}
	go func(g *Group, m int) {
		h := g.send(g.render(g.getOnPassTemplate(), u), m)
		time.Sleep(g.getTimeout())
		if g.existsf("verbose") {
			return
		}
		go g.delete(m)
		go g.delete(h)
	}(g, m.ID)
}

func (g *Group) onFail(m *telebot.Message, u *telebot.User) {
	log.Tracef("group(%d).onfail(uid=%d)", g.id, u.ID)
	g.delf("user:%d:pending", u.ID)
	if !g.existsf("verbose") && m != nil {
		go g.delete(m.ID)
	}
	switch g.getFailAction() {
	case actionKick:
		go g.kick(u.ID)
	case actionMute:
		go g.mute(u.ID)
	case actionBan:
		go g.ban(u.ID)
	}

	if g.existsf("quiet") {
		return
	}
	go func(g *Group) {
		h := g.send(g.render(g.getOnFailTemplate(), u), 0)
		time.Sleep(g.getTimeout())
		if !g.existsf("verbose") {
			g.delete(h)
		}
	}(g)
}

func (g *Group) handleCommand(m *telebot.Message) {
	cmd, arg := bot.ParseCommand(m)
	switch cmd {

	case "ping":
		go g.send(g.format("ping.pong", time.Since(m.Time()).Round(time.Millisecond)), m.ID)

	case "help":

	case "status":
		if !g.checkIsAdmin(m) {
			return
		}
		if g.existsf("enabled") {
			go g.send(g.format("status.enable"), m.ID)
		} else {
			go g.send(g.format("status.disable"), m.ID)
		}

	case "enable":
		if !g.checkIsAdmin(m) {
			return
		}
		g.setf("enabled", "true", 0)
		go g.send(g.format("status.enable"), m.ID)

	case "disable":
		if !g.checkIsAdmin(m) {
			return
		}
		g.delf("enabled")
		go g.send(g.format("status.disable"), m.ID)

	case "action":
		if !g.checkIsAdmin(m) {
			return
		}
		if len(arg) > 0 {
			a := action(arg)
			if a != actionKick && a != actionMute && a != actionBan {
				go g.send(g.format("cmd.bad_param")+"\n\n"+g.format("action.help.full"), m.ID)
				return
			}
			g.setf("action", arg, 0)
		}
		v := g.format("action." + string(g.getFailAction()))
		go g.send(g.format("action.query", v), m.ID)

	case "timeout":
		if !g.checkIsAdmin(m) {
			return
		}
		if len(arg) > 0 {
			x, err := strconv.Atoi(arg)
			if err != nil || x <= 0 || x >= 2147483648 {
				go g.send(g.format("cmd.bad_param")+"\n\n"+g.format("timeout.help.full"), m.ID)
				return
			}
			g.setf("timeout", arg, 0)
		}
		x := int(g.getTimeout().Seconds())
		s := g.format("timeout.query", x)
		if x < 10 {
			s += "\n\n" + g.format("timeout.notice")
		}
		go g.send(s, m.ID)

	case "lang":
		if !g.checkIsAdmin(m) {
			return
		}
		if len(arg) > 0 {
			g.setf("lang", arg, 0)
		}
		l, err := g.getf("lang")
		if err == redis.Nil {
			l = "en_US"
		}
		go g.send(g.format("lang.query", l, i18n.AllLanguages()), m.ID)

	case "verbose":
		if !g.checkIsAdmin(m) {
			return
		}
		if len(arg) > 0 {
			switch arg {
			case "on":
				g.setf("verbose", "true", 0)
				g.delf("quiet")
			case "off":
				g.delf("verbose")
			default:
				go g.send(g.format("cmd.bad_param")+"\n\n"+g.format("verbose.help.full"), m.ID)
				return
			}
		}
		if g.existsf("verbose") {
			go g.send(g.format("verbose.on"), m.ID)
		} else {
			go g.send(g.format("verbose.off"), m.ID)
		}

	case "quiet":
		if !g.checkIsAdmin(m) {
			return
		}
		if len(arg) > 0 {
			switch arg {
			case "on":
				g.setf("quiet", "true", 0)
				g.delf("verbose")
			case "off":
				g.delf("quiet")
			default:
				go g.send(g.format("cmd.bad_param")+"\n\n"+g.format("quiet.help.full"), m.ID)
				return
			}
		}
		if g.existsf("quiet") {
			go g.send(g.format("quiet.on"), m.ID)
		} else {
			go g.send(g.format("quiet.off"), m.ID)
		}

	case "onjoin":
		if !g.checkIsAdmin(m) {
			return
		}
		if len(arg) > 0 {
			g.setf("onjoin:template", arg, 0)
		}
		go g.send(g.format("onjoin.query", g.getOnJoinTemplate()), m.ID)

	case "onpass":
		if !g.checkIsAdmin(m) {
			return
		}
		if len(arg) > 0 {
			g.setf("onpass:template", arg, 0)
		}
		go g.send(g.format("onpass.query", g.getOnPassTemplate()), m.ID)

	case "onfail":
		if !g.checkIsAdmin(m) {
			return
		}
		if len(arg) > 0 {
			g.setf("onfail:template", arg, 0)
		}
		go g.send(g.format("onfail.query", g.getOnFailTemplate()), m.ID)

	case "refresh":
		u := m.Sender.ID
		if m.ReplyTo != nil {
			u = m.ReplyTo.Sender.ID
		}
		g.delf("user:%d:role", u)
		go g.delete(m.ID)

	case "reverify":
		if !g.checkIsAdmin(m) || !g.checkHasReply(m) || !g.checkNoOp(m) {
			return
		}
		if m.ReplyTo.IsService() {
			for _, u := range m.ReplyTo.UsersJoined {
				g.onJoin(m, &u)
			}
		} else {
			g.onJoin(m, m.ReplyTo.Sender)
		}

	case "pass":
		if !g.checkIsAdmin(m) || !g.checkHasReply(m) || !g.checkNoOp(m) {
			return
		}
		if m.ReplyTo.IsService() {
			for _, u := range m.ReplyTo.UsersJoined {
				g.onPass(m, &u)
			}
		} else {
			g.onPass(m, m.ReplyTo.Sender)
		}

	case "fail":
		if !g.checkIsAdmin(m) || !g.checkHasReply(m) || !g.checkNoOp(m) {
			return
		}
		if m.ReplyTo.IsService() {
			for _, u := range m.ReplyTo.UsersJoined {
				g.onFail(nil, &u)
			}
		} else {
			g.onFail(nil, m.ReplyTo.Sender)
		}

	}
}

func (g *Group) checkIsAdmin(m *telebot.Message) bool {
	if g.getRole(m.Sender.ID) == roleAdmin {
		return true
	}
	if g.existsf("quiet") {
		go g.delete(m.ID)
		return false
	}
	go func(g *Group, m int) {
		h := g.send(g.format("cmd.not_admin"), m)
		if g.existsf("verbose") {
			return
		}
		time.Sleep(g.getTimeout())
		go g.delete(m)
		go g.delete(h)
	}(g, m.ID)
	return false
}

func (g *Group) checkHasReply(m *telebot.Message) bool {
	if m.ReplyTo != nil {
		return true
	}
	if g.existsf("quiet") {
		go g.delete(m.ID)
		return false
	}
	go func(g *Group, m int) {
		h := g.send(g.format("cmd.need_reply"), m)
		if g.existsf("verbose") {
			return
		}
		time.Sleep(g.getTimeout())
		go g.delete(m)
		go g.delete(h)
	}(g, m.ID)
	return false
}

func (g *Group) checkNoOp(m *telebot.Message) bool {
	if g.existsf("enabled") {
		if m.ReplyTo.IsService() {
			if len(m.ReplyTo.UsersJoined) > 0 {
				return true
			}
		} else {
			return true
		}
	}
	if g.existsf("quiet") {
		go g.delete(m.ID)
		return false
	}
	go func(g *Group, m int) {
		h := g.send(g.format("cmd.no_op"), m)
		if g.existsf("verbose") {
			return
		}
		time.Sleep(g.getTimeout())
		go g.delete(m)
		go g.delete(h)
	}(g, m.ID)
	return false
}

func (g *Group) getRole(u int) role {
	ttl := 120 * time.Second
	r, err := g.getf("user:%d:role", u)
	if err == redis.Nil {
		e := bot.GetChatMember(g.id, u)
		switch {
		case e == nil:
			r = string(roleNone)
		case e.Role == telebot.Creator || e.CanRestrictMembers:
			r = string(roleAdmin)
		default:
			r = string(roleMember)
		}
		g.setf("user:%d:role", r, ttl, u)
	}
	return role(r)
}

var templateRe = regexp.MustCompile("\\$.")

func (g *Group) render(template string, u *telebot.User) string {
	template = bot.EscapeHTML(template)
	return templateRe.ReplaceAllStringFunc(template, func(s string) string {
		switch s[1] {
		case '$':
			return "$"
		case 'u':
			n := u.FirstName
			if len(u.LastName) > 0 {
				n = n + " " + u.LastName
			}
			return fmt.Sprintf(`<a href="tg://user?id=%d">%s</a>`, u.ID, bot.EscapeHTML(n))
		case 't':
			return strconv.Itoa(int(g.getTimeout().Seconds()))
		default:
			return ""
		}
	})
}

func (g *Group) getOnJoinTemplate() string {
	fallback := g.format("onjoin.default")
	s, err := g.getf("onjoin:template")
	if err == redis.Nil {
		return fallback
	}
	return s
}

func (g *Group) getOnPassTemplate() string {
	fallback := g.format("onpass.default")
	s, err := g.getf("onpass:template")
	if err == redis.Nil {
		return fallback
	}
	return s
}

func (g *Group) getOnFailTemplate() string {
	fallback := g.format("onfail.default")
	s, err := g.getf("onfail:template")
	if err == redis.Nil {
		return fallback
	}
	return s
}

func (g *Group) getFailAction() action {
	fallback := actionKick
	s, err := g.getf("action")
	if err == redis.Nil {
		return fallback
	}
	return action(s)
}

func (g *Group) getTimeout() time.Duration {
	fallback := 60 * time.Second
	s, err := g.getf("timeout")
	if err == redis.Nil {
		return fallback
	}
	x, _ := strconv.Atoi(s)
	return time.Duration(x) * time.Second
}

func (g *Group) getLang() string {
	fallback := "en_US"
	s, err := g.getf("lang")
	if err == redis.Nil {
		return fallback
	}
	return s
}

func (g *Group) getf(f string, args ...interface{}) (string, error) {
	k := fmt.Sprintf("group:%d:", g.id) + fmt.Sprintf(f, args...)
	return redis.Get(k)
}

func (g *Group) setf(f string, v string, ttl time.Duration, args ...interface{}) {
	k := fmt.Sprintf("group:%d:", g.id) + fmt.Sprintf(f, args...)
	redis.Set(k, v, ttl)
}

func (g *Group) delf(f string, args ...interface{}) {
	k := fmt.Sprintf("group:%d:", g.id) + fmt.Sprintf(f, args...)
	redis.Del(k)
}

func (g *Group) existsf(f string, args ...interface{}) bool {
	k := fmt.Sprintf("group:%d:", g.id) + fmt.Sprintf(f, args...)
	r, _ := redis.Exists(k)
	return r
}

func (g *Group) send(html string, reply int) int {
	return bot.Send(g.id, html, reply)
}

func (g *Group) delete(m int) {
	bot.Delete(g.id, m)
}

func (g *Group) mute(u int) {
	bot.Mute(g.id, u)
}

func (g *Group) ban(u int) {
	g.delf("user:%d:role", u)
	bot.Ban(g.id, u)
}

func (g *Group) unban(u int) {
	bot.Unban(g.id, u)
}

func (g *Group) kick(u int) {
	g.ban(u)
	g.unban(u)
}

func (g *Group) format(key string, args ...interface{}) string {
	return i18n.Format(g.getLang(), key, args...)
}

File: group.ts
import "source-map-support/register.js";

import TelegramBot from "node-telegram-bot-api";

import * as bot from "./bot.js";
import * as config from "./config.js";
import * as i18n from "./i18n/index.js";
import Mutex from "./mutex.js";
import * as redis from "./redis.js";
import { logger } from "./log.js";

const botLogger = logger.child({ scope: "bot" });
const groupLogger = logger.child({ scope: "group" });

type Role = "none" | "member" | "admin";

type Action = "kick" | "mute" | "ban";

export default class Group {

    private static index = new Map<number, Group>();

    public static get(id: number): Group {
        let g = Group.index.get(id);
        if (g !== undefined) {
            groupLogger.silly("get()", { chat: id, status: "loaded" });
            return g;
        }
        g = new Group(id);
        Group.index.set(id, g);
        groupLogger.silly("get()", { chat: id, status: "stored" });
        return g;
    }

    public readonly id: number;

    private readonly resolvers: Map<number, (passed: boolean) => void>;

    private readonly mutex: Mutex;

    public constructor(id: number) {
        this.id = id;
        this.resolvers = new Map();
        this.mutex = new Mutex();
    }

    public async handleMessage(m: TelegramBot.Message): Promise<void> {
        botLogger.verbose("update", { chat: m.chat.id, msg: m.message_id });
        for (const u of (m.new_chat_members || [])) {
            await this.delKey(`user:${u.id}:role`);
            if (u.id === bot.getMe().id) {
                await this.delKey("enabled");
            }
        }
        if (await this.handleVerification(m)) {
            return;
        }
        if (await this.handleCommand(m)) {
            return;
        }
    }

    private async handleVerification(m: TelegramBot.Message): Promise<boolean> {
        if (await this.existsKey(`user:${m.from?.id}:pending`)) {
            if (m.sticker !== undefined) {
                await this.onPass(m, m.from as TelegramBot.User);
            } else {
                await this.delMsg(m.message_id);
            }
            return true;
        }
        if (!await this.existsKey("enabled")) {
            return false;
        }
        if (await this.getRole(bot.getMe().id) !== "admin") {
            await this.delKey("enabled");
            await this.send(await this.format("bot.angry"));
            await this.leave();
            return true;
        }
        if (m.new_chat_members !== undefined) {
            await Promise.all(m.new_chat_members.map((u) => this.onJoin(m, u)));
            return true;
        }
        return false;
    }

    private async onJoin(msg: TelegramBot.Message, user: TelegramBot.User): Promise<void> {
        if (await this.existsKey(`user:${user.id}:pending`)) {
            groupLogger.info("onjoin", { chat: this.id, msg: msg.message_id, user: user.id, status: "dup" });
            return;
        }

        groupLogger.info("onjoin", { chat: this.id, msg: msg.message_id, user: user.id, status: "start" });
        await this.setKey(`user:${user.id}:pending`, "true");
        const h = await this.send(await this.render(await this.getTemplate("onjoin"), user), msg.message_id);

        const passed = await Promise.race([
            this.sleep(false),
            new Promise<boolean>(async (resolve) => {
                const resolver = (passed: boolean): void => {
                    this.resolvers.delete(user.id);
                    resolve(passed);
                };
                this.resolvers.set(user.id, resolver);
                if (user.id === bot.getMe().id) {
                    try {
                        const s = "CAACAgUAAxkBAAEI_IFgKqYpeH28bSvB_qd3ybC5vS-RxwACsgADVl_YH824--1Q953HHgQ";
                        const w = await bot.getAPI().sendSticker(this.id, s);
                        await this.onPass(w, w.from as TelegramBot.User);
                    } catch { }
                }
            }),
        ]);

        if (!await this.existsKey("verbose")) {
            await this.delMsg(h);
        }
        if (passed) {
            return;
        }

        if (!await this.existsKey("verbose")) {
            await this.delMsg(msg.message_id);
        }
        await this.onFail(user);
    }

    private async onPass(msg: TelegramBot.Message, user: TelegramBot.User): Promise<void> {
        if (!await this.existsKey(`user:${user.id}:pending`)) {
            return;
        }
        groupLogger.info("onpass", { chat: this.id, msg: msg.message_id, user: user.id });
        await this.delKey(`user:${user.id}:pending`);
        const resolve = this.resolvers.get(user.id);
        if (resolve !== undefined) {
            resolve(true);
        }
        if (await this.existsKey("quiet")) {
            await this.delMsg(msg.message_id);
            return;
        }
        await this.send(await this.render(await this.getTemplate("onpass"), user), msg.message_id);
    }

    private async onFail(user: TelegramBot.User): Promise<void> {
        if (!await this.existsKey(`user:${user.id}:pending`)) {
            return;
        }
        groupLogger.info("onpass", { chat: this.id, user: user.id });
        await this.delKey(`user:${user.id}:pending`);
        const resolve = this.resolvers.get(user.id);
        if (resolve !== undefined) {
            resolve(false);
        }
        await this[await this.getAction()](user.id);
        if (await this.existsKey("quiet")) {
            return;
        }
        const f = await this.send(await this.render(await this.getTemplate("onfail"), user));
        if (await this.existsKey("verbose")) {
            return;
        }
        await this.sleep();
        await this.delMsg(f);
    }

    private async handleCommand(m: TelegramBot.Message): Promise<boolean> {
        const [cmd, arg] = bot.parseCommand(m);
        switch (cmd) {
            case "start":
            case "help":
                if (!await this.checkFromAdmin(m, true)) {
                    break;
                }
                const help: ("" | i18n.TranslationKey)[] = [
                    "sticker_captcha_bot.help",
                    "",
                    "help.help",
                    "ping.help",
                    "refresh.help",
                    "",
                    "status.help",
                    "enable.help",
                    "disable.help",
                    "",
                    "lang.help",
                    "timeout.help",
                    "action.help",
                    "",
                    "onjoin.help",
                    "onpass.help",
                    "onfail.help",
                    "template.help",
                    "",
                    "verbose.help",
                    "quiet.help",
                    "",
                    "reverify.help",
                    "pass.help",
                    "fail.help",
                    "",
                    "open_source.help",
                ];
                const lang = await this.getLang()
                await this.send(help.map((l: "" | i18n.TranslationKey): string => {
                    return l === "" ? "" : i18n.format(lang, l);
                }).join("\n"), m.message_id);
                break;

            case "ping":
                const l = (Math.ceil(Date.now() / 1e3) - m.date).toString() + "s";
                await this.send(await this.format("ping.pong", l), m.message_id);
                break;

            case "refresh":
                let u = m.from?.id;
                if (m.reply_to_message !== undefined) {
                    u = m.reply_to_message.from?.id;
                }
                if (u !== undefined) {
                    await this.delKey(`user:${u}:role`);
                }
                await this.delMsg(m.message_id);
                break;

            case "status":
                if (!await this.checkFromAdmin(m)) {
                    break;
                }
                const status = await this.existsKey("enabled") ? "enable" : "disable";
                await this.send(await this.format(`status.${status}`), m.message_id);
                break;

            case "enable":
            case "disable":
                if (!await this.checkFromAdmin(m)) {
                    break;
                }
                if (cmd === "enable") {
                    if (await this.getRole(bot.getMe().id, true) !== "admin") {
                        await this.send(await this.format("bot.not_admin"), m.message_id);
                        break;
                    }
                    await this.setKey("enabled", "true");
                } else {
                    await this.delKey("enabled");
                }
                await this.send(await this.format(`status.${cmd}`), m.message_id);
                break;

            case "lang":
                if (!await this.checkFromAdmin(m, true)) {
                    break;
                }
                if (arg !== undefined) {
                    await this.setKey("lang", arg);
                }
                await this.send(await this.format("lang.query", await this.getLang(), i18n.allLangs()), m.message_id);
                break;

            case "action":
                if (!await this.checkFromAdmin(m)) {
                    break;
                }
                if (arg !== undefined) {
                    if (!["kick", "mute", "ban"].includes(arg)) {
                        await this.send(await this.format("cmd.bad_param"), m.message_id);
                        break;
                    }
                    await this.setKey("action", arg);
                }
                const v = await this.format(`action.${await this.getAction()}`);
                await this.send(await this.format("action.query", v), m.message_id);
                break;

            case "timeout":
                if (!await this.checkFromAdmin(m)) {
                    break;
                }
                if (arg !== undefined) {
                    const x = Number.parseInt(arg);
                    if (Number.isNaN(x) || x <= 0 || x >= 2147483.648) {
                        await this.send(await this.format("cmd.bad_param"), m.message_id);
                        break;
                    }
                    await this.setKey("timeout", x.toString());
                }
                const x = await this.getTimeout();
                let s = await this.format("timeout.query", x);
                if (x < 10) {
                    s += "\n\n" + await this.format("timeout.notice");
                }
                await this.send(s, m.message_id);
                break;

            case "onjoin":
            case "onpass":
            case "onfail":
                if (!await this.checkFromAdmin(m)) {
                    break;
                }
                if (arg !== undefined) {
                    await this.setKey(`${cmd}:template`, arg);
                }
                await this.send(await this.format(`${cmd}.query`, await this.getTemplate(cmd)), m.message_id)
                break;

            case "verbose":
            case "quiet":
            case "debug":
                if (!await this.checkFromAdmin(m)) {
                    break;
                }
                switch (arg) {
                    case "on":
                        await this.setKey(cmd, "true");
                        const conflict = {
                            "verbose": "quiet",
                            "quiet": "verbose",
                            "debug": undefined,
                        }[cmd];
                        if (conflict !== undefined) {
                            await this.delKey(conflict);
                        }
                        await this.send(await this.format(`${cmd}.on`), m.message_id);
                        break;
                    case "off":
                        await this.delKey(cmd);
                        await this.send(await this.format(`${cmd}.off`), m.message_id);
                        break;
                    case undefined:
                        const status = await this.existsKey(cmd) ? "on" : "off";
                        await this.send(await this.format(`${cmd}.${status}`), m.message_id);
                        break;
                    default:
                        await this.send(await this.format("cmd.bad_param"), m.message_id);
                }
                break;

            case "reverify":
            case "pass":
            case "fail":
                if (!await this.checkFromAdmin(m) || !await this.checkHasReply(m)) {
                    break;
                }
                const rep = m.reply_to_message as TelegramBot.Message;
                const func = {
                    reverify: (u: TelegramBot.User) => this.onJoin(rep, u),
                    pass: (u: TelegramBot.User) => this.onPass(m, u),
                    fail: (u: TelegramBot.User) => this.onFail(u),
                }[cmd];
                if (rep.new_chat_members !== undefined) {
                    await Promise.all(rep.new_chat_members.map((u) => func(u)));
                } else {
                    await func(rep.from as TelegramBot.User);
                }
                break;

            case "id":
                await this.send(`<code>${this.id}</code>`, m.message_id);
                break;

            default:
                return false;
        }

        return true;
    }

    private async checkFromAdmin(m: TelegramBot.Message, allowPrivate?: boolean): Promise<boolean> {
        if (m.chat.type === "private") {
            if (allowPrivate) {
                return true;
            }
            await this.send(await this.format("cmd.not_in_group"), m.message_id);
            return false
        }
        if (await this.getRole(m.from?.id as number) !== "admin") {
            await this.send(await this.format("cmd.not_admin"), m.message_id);
            return false;
        }
        return true;
    }

    private async checkHasReply(m: TelegramBot.Message): Promise<boolean> {
        if (m.reply_to_message !== undefined) {
            return true;
        }
        await this.send(await this.format("cmd.need_reply"), m.message_id);
        return false;
    }

    private async sleep(): Promise<void>;
    private async sleep<T>(res: T): Promise<T>;
    private async sleep(res?: any): Promise<any> {
        const time = await this.getTimeout() * 1e3;
        return new Promise((resolve) => setTimeout(() => resolve(res), time));
    }

    private async getRole(user: number, refresh: boolean = false): Promise<Role> {
        let r: string | undefined;
        if (!refresh) {
            r = await this.getKey(`user:${user}:role`);
            if (r !== undefined) {
                // TODO: type check
                return r as Role;
            }
        }
        let e;
        while (true) {
            try {
                e = await bot.getChatMember(this.id, user, await this.existsKey("debug"));
                break;
            } catch { }
        }
        if (e === undefined || e.status === "kicked" || e.status == "left") {
            r = "none";
        } else if (e.status === "creator" || e.can_restrict_members) {
            r = "admin";
        } else {
            r = "member";
        }
        await this.setKey(`user:${user}:role`, r, 120);
        return r as Role;
    }

    private async render(tmpl: string, user: TelegramBot.User): Promise<string> {
        tmpl = bot.escapeHTML(tmpl);
        let res = "";
        for (let i = 0; i < tmpl.length; i++) {
            if (tmpl[i] !== "$") {
                res += tmpl[i];
                continue;
            }
            i++;
            switch (tmpl[i]) {
                case "$":
                    res += "$";
                    break;
                case "u":
                    let n = user.first_name;
                    if (user.last_name !== undefined) {
                        n = `${n} ${user.last_name}`;
                    }
                    res += `<a href="tg://user?id=${user.id}">${bot.escapeHTML(n)}</a>`;
                    break;
                case "i":
                    res += `<a href="tg://user?id=${user.id}">${user.id}</a>`;
                    break;
                case "t":
                    res += (await this.getTimeout()).toString();
                    break;
                default:
                // no-op;
            }
        }
        return res;
    }

    private async getTemplate(on: "onjoin" | "onpass" | "onfail"): Promise<string> {
        const s = await this.getKey(`${on}:template`);
        return s !== undefined ? s : this.format(`${on}.default`);
    }

    private async getAction(): Promise<Action> {
        const s = await this.getKey("action");
        if (s === undefined) {
            return config.get("default_action", "kick");
        }
        // TODO: type check
        return s as Action;
    }

    private async getTimeout(): Promise<number> {
        const s = await this.getKey("timeout");
        const t = Number.parseInt(s as string);
        return Number.isNaN(t) ? config.get("default_timeout", 60) : t;
    }

    private async getLang(): Promise<string> {
        return await this.getKey("lang") || config.get("default_lang", "en_US");
    }

    private async getKey(key: string): Promise<string | undefined> {
        return redis.get(`group:${this.id}:${key}`);
    }

    private async setKey(key: string, val: string, ttl?: number): Promise<void> {
        return redis.set(`group:${this.id}:${key}`, val, ttl);
    }

    private async delKey(key: string): Promise<void> {
        return redis.del(`group:${this.id}:${key}`);
    }

    private async existsKey(key: string): Promise<boolean> {
        return redis.exists(`group:${this.id}:${key}`);
    }

    private async send(html: string, reply?: number): Promise<number> {
        return bot.send(this.id, html, await this.existsKey("debug"), reply);
    }

    private async delMsg(msg: number): Promise<boolean> {
        return bot.del(this.id, msg);
    }

    private async mute(user: number): Promise<boolean> {
        return bot.mute(this.id, user);
    }

    private async ban(user: number): Promise<void> {
        await bot.ban(this.id, user);
        await this.delKey(`user:${user}:role`);
    }

    private async kick(user: number): Promise<void> {
        await bot.ban(this.id, user);
        await this.delKey(`user:${user}:role`);
        await bot.unban(this.id, user);
    }

    private async leave(): Promise<boolean> {
        return bot.leaveChat(this.id);
    }

    private async format(key: i18n.TranslationKey, ...args: any[]): Promise<string> {
        return i18n.format(await this.getLang(), key, ...args);
    }

};
